kymatio/scattering2d/backend/torch_backend.py

```diff
@@ -0,0 +1,320 @@
+        if pad_size_tmp[0] == self.input_size[0]:
+            pad_size_tmp[0] -= 1
+            pad_size_tmp[1] -= 1
+        if pad_size_tmp[2] == self.input_size[1]:
+            pad_size_tmp[2] -= 1
+            pad_size_tmp[3] -= 1
+        # Pytorch expects its padding as [left, right, top, bottom]
+        self.padding_module = ReflectionPad2d([pad_size_tmp[2], pad_size_tmp[3],
+                                               pad_size_tmp[0], pad_size_tmp[1]])
+
+    def __call__(self, x):
+        """Applies padding and maps to complex.
+            
+            Parameters
+            ----------
+            x : tensor
+                Real tensor input to be padded and sent to complex domain.
+
+            Returns
+            -------
+            output : tensor
+                Complex torch tensor that has been padded.
+
+        """
+        batch_shape = x.shape[:-2]
+        signal_shape = x.shape[-2:]
+        x = x.reshape((-1, 1) + signal_shape)
+        if not self.pre_pad:
+            x = self.padding_module(x)
+
+            # Note: PyTorch is not effective to pad signals of size N-1 with N
+            # elements, thus we had to add this fix.
+            if self.pad_size[0] == self.input_size[0]:
+                x = torch.cat([x[:, :, 1, :].unsqueeze(2), x, x[:, :, x.shape[2] - 2, :].unsqueeze(2)], 2)
+            if self.pad_size[2] == self.input_size[1]:
+                x = torch.cat([x[:, :, :, 1].unsqueeze(3), x, x[:, :, :, x.shape[3] - 2].unsqueeze(3)], 3)
+
+        output = x.new_zeros(x.shape + (2,))
+        output[..., 0] = x
+        output = output.reshape(batch_shape + output.shape[-3:])
+        return output
+
+def unpad(in_):
+    """Unpads input.
+        
+        Slices the input tensor at indices between 1:-1.
+
+        Parameters
+        ----------
+        in_ : tensor
+            Input tensor.
+
+        Returns
+        -------
+        in_[..., 1:-1, 1:-1] : tensor
+            Output tensor.  Unpadded input.
+
+    """
+    return torch.unsqueeze(in_[..., 1:-1, 1:-1], -3)
+
+class SubsampleFourier(object):
+    """Subsampling of a 2D image performed in the Fourier domain
+        
+        Subsampling in the spatial domain amounts to periodization
+        in the Fourier domain, hence the formula.
+
+        Parameters
+        ----------
+        x : tensor
+            Input tensor with at least 5 dimensions, the last being the real
+            and imaginary parts.
+        k : int
+            Integer such that x is subsampled by 2**k along the spatial variables.
+
+        Returns
+        -------
+        out : tensor
+            Tensor such that its Fourier transform is the Fourier
+            transform of a subsampled version of x, i.e. in
+            F^{-1}(out)[u1, u2] = F^{-1}(x)[u1 * (2**k), u2 * (2**k)].
+
+    """
+    def __call__(self, x, k):
+        if not x.is_contiguous():
+            raise RuntimeError('Input should be contiguous.')
```

Comment:
Do we want this contiguous check here? I added it bc the skcuda backend had it, and changed a test correspondingly, but not sure if we want this.

Reply:
Yes. In general, we want the backends to have the same interface, which means they have the same errors raised under the same circumstances (as much as possible).